[docs] Create platform-specific-modules guide for router (#22846)
expo/expo: docs/pages/router/advance/platform-specific-modules.mdx

---
title: Platform-specific Modules
description: Learn how to switch modules based on the platform in Expo Router.
hideTOC: true
---

import { FileTree } from '~/ui/components/FileTree';

> While building your app, you may want to show specific content based on the current platform. Platform-specific modules can be used to make the experience more native to a given platform. There are a few ways to achieve this in Expo Router.

Reviewer: ```suggestion
While building your app, you may want to show specific content based on the current platform. Platform-specific modules can make the experience more native to a given platform. The following sections describe the ways you can achieve this with Expo Router.
```

### Platform module

You can use the [`Platform`](https://reactnative.dev/docs/platform-specific-code#platform-module) module from React Native to detect the current platform and render the appropriate content based on the result. For example, you can render a `Tabs` layout on native and a custom layout on the web.

```js app/_layout.js
import { Platform } from 'react-native';
import { Link, Tabs } from 'expo-router';

export default function Layout() {
  if (Platform.OS === 'web') {
    // Use a basic custom layout on web.
    return (
      <div style={{ flex: 1 }}>
        <header>
          <Link href="/">Home</Link>
          <Link href="/settings">Settings</Link>
        </header>
        <Slot />
      </div>
    );
  }
  // Use a native bottom tabs layout on native platforms.
  return (
    <Tabs>
      <Tabs.Screen name="index" options={{ title: 'Home' }} />
      <Tabs.Screen name="settings" options={{ title: 'Settings' }} />
    </Tabs>
  );
}
```

### Platform specific extensions

> Metro bundler's platform-specific extensions (e.g. `.ios.js`, `.native.ts`) are not supported in the app directory, this ensures that routes are universal across platforms for deep linking. However, you can create platform-specific files outside of the `/app` directory, and use them from within the `/app` directory.

Reviewer: ```suggestion
Metro bundler's platform-specific extensions (for example, **.ios.js** or **.native.ts**) are not supported in the **app** directory. This ensures that routes are universal across platforms for deep linking. However, you can create platform-specific files outside the **app** directory and use them from within the **app** directory.
```

Consider the following project:

<FileTree
  files={[
    'app/_layout.js',
    'app/index.js',
    'app/about.js',
    'components/about.js',
    'components/about.ios.js',
    'components.web.js',
  ]}
/>

> Suppose the designs require you to build different `about` screens for each platform. In that case, you can create a component for each platform in the `components/` directory using platform extensions. When imported, Metro will ensure the correct version of the component is used based on the current platform. You can then re-export the the component as a screen in the app directory.

Reviewer: ```suggestion
For example, the designs require you to build different `about` screens for each platform. In that case, you can create a component for each platform in the **components** directory using platform extensions. When imported, Metro will ensure the correct component version is used based on the current platform. You can then re-export the the component as a screen in the **app** directory.
```

```js app/about.js
export { default } from '../components/about';
```

> > **Best Practice:** Always provide a file without a platform extension to ensure every platform has a default implementation.

Reviewer: ```suggestion
>  **info** **Best practice:** Always provide a file without a platform extension to ensure every platform has a default implementation.
```
